### ISSUE TITLE
native Spectrum: CPU-streaming SNR layer scan + targeted unfreeze (soup spectrum scan)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#266]
## Context

Soup can already attach the **upstream** Spectrum package as a trainer plugin (`utils/trainer_plugins.py` catalog entry since v0.45.0, live instantiation since v0.53.7 #105). But that is a wrapper: without the external package there is no feature, and Soup has no native SNR scanner or unfrozen-config generator.

Spectrum (arXiv:2406.06623, Apache-2.0 — license-compatible for porting) scans each layer's weight Signal-to-Noise Ratio (SVD spectrum vs the Marchenko-Pastur noise threshold) and trains only the top-n% highest-SNR layers — full-FT quality at a fraction of the trainable parameters.

**Our edge over the original:** SNR is computed from the *weights*, not activations — so the scan can stream safetensors shards tensor-by-tensor through numpy with NO model load and NO GPU. The reference implementation needs an 8xH100 node to scan a 70B model; a native Soup scanner can do it on a laptop CPU. That is a headline feature for the v0.72 Colab notebooks ("scan a 70B's layer SNR on CPU, fine-tune only the top 50%").

## Proposed fix path

1. **`soup_cli/utils/spectrum_scan.py`** — pure-numpy SNR kernel:
   - `compute_layer_snr(matrix) -> float` (SVD singular values; noise threshold via Marchenko-Pastur distribution per the paper; SNR = signal energy above threshold / noise energy below).
   - `scan_model_snr(model_dir_or_repo, *, module_filter)` — iterate safetensors shards via `safetensors.safe_open` (lazy, per-tensor — never materialise the full checkpoint), default module filter = MLP + attention projections (`q/k/v/o_proj`, `gate/up/down_proj`), Conv1D (GPT-2) transpose-aware like v0.71.16 `edit_kernels`.
   - `select_top_percent(snr_results, top_percent) -> tuple[str, ...]` — sorted module names.
   - Scan cache at `~/.soup/spectrum/<model-slug>.json` (env override containment-checked, mirrors `SOUP_BATCH_CACHE_PATH` policy).
2. **`soup spectrum scan --model <id|path> --top-percent 50 [--output soup-patch.yaml] [--modules mlp,attn|all]`** — new CLI: renders the SNR table + emits a ready-to-paste `training.unfrozen_parameters` YAML block.
3. **Schema:** `TrainingConfig.unfrozen_parameters: Optional[List[str]]` (regex-or-exact module patterns; caps + null-byte rejection per project validator policy). Cross-validators: mutually exclusive with `lora` (full-FT-style method), with `freeze_layers`/`freeze_ratio`, and with `train_router_only`.
4. **Trainer wiring:** shared helper `apply_unfrozen_parameters(model, patterns)` (freeze all -> unfreeze matching; returns trainable-param count) in `utils/freeze.py` or a sibling, wired through the SFT-family trainers BEFORE the optimizer is built (matches the LLaMA Pro `apply_block_expansion_if_configured` centralisation pattern).
5. The existing `spectrum` trainer-plugin wrapper stays untouched (back-compat).

## Acceptance criteria

- [ ] `compute_layer_snr` matches the paper's Marchenko-Pastur formulation; unit-tested on synthetic matrices (pure-signal high SNR, gaussian-noise low SNR)
- [ ] `soup spectrum scan` streams shards without loading the model (no `from_pretrained`); peak RSS bounded by largest single tensor
- [ ] Scan of SmolLM2-135M completes on CPU; live smoke trains with top-25% unfrozen and loss decreases (RTX 3050)
- [ ] `training.unfrozen_parameters` schema field + cross-validators (lora / freeze_layers / train_router_only exclusivity) with named error messages
- [ ] Generated YAML block round-trips through `load_config_from_string`
- [ ] Scan cache hit short-circuits a re-scan; cache path containment-checked
- [ ] Tests follow project policy (happy + failure + bounds + symlink/TOCTOU on cache + no top-level torch in the scan util)

**Comment by MakazhanAlpamys** [comment]
Shipped in **v0.71.23**. See the [release notes](https://github.com/MakazhanAlpamys/Soup/releases/tag/v0.71.23).

Try it:
```bash
soup spectrum scan --model HuggingFaceTB/SmolLM2-135M --top-percent 50 --modules mlp,attn -o patch.yaml
```
Then paste the `training.unfrozen_parameters` block from `patch.yaml` into your `soup.yaml` (with `quantization: none`) and run `soup train` — it full-fine-tunes only the high-SNR layers. The scan streams safetensors per-tensor (no model load), so it runs on a CPU box.

LISA (per-step layerwise importance sampling) is the follow-up in #267.
